maintenance/wikia/uncycloUsersMigrator.php

```diff
@@ -0,0 +1,695 @@
+		$whoIs = User::whoIs( $newUserId );
+
+		if ( $whoIs !== false ) {
+			$this->output( sprintf( "\nCan't change the ID of %s - clashes with %s!\n", $user->getName(), $whoIs ) );
+
+			if (!$this->isDryRun) {
+				throw new UncycloUsersMigratorException(sprintf('IDs clash for %s (new ID taken from shared DB would be #%d)', $user->getName(), $newUserId));
+			}
+		}
+
+		if ( $this->isDryRun ) {
+			return;
+		}
+
+		// update user_id in MW tables
+		$this->updateTables( self::UPDATE_TABLE_USER_ID, $user->getId(), $newUserId );
+	}
+
+	/**
+	 * Rename uncyclopedia user
+	 *
+	 * Updates fields with user name only!
+	 *
+	 * @param User $uncycloUser
+	 * @return User user object with a changed name
+	 */
+	protected function doRenameUncycloUser( User $uncycloUser ) {
+		// don't touch uncyclo users in "global users only" mode
+		if ( $this->onlyRenameGlobalUsers ) {
+			return $uncycloUser;
+		}
+
+		$user = clone $uncycloUser;
+		$newName = self::PREFIX_RENAME_UNCYCLO . $user->getName();
+
+		// check user name conflicts
+		if ( $this->getGlobalUserByName( $newName ) ) {
+			$newName .= '-1';
+		}
+
+		$this->output( sprintf('> renaming uncyclo user to "%s"...', $newName) );
+
+		$this->renamedUnclycloAccounts++;
+
+		if ( !$this->isDryRun ) {
+			// update user_name in MW tables
+			$this->updateTables(self::UPDATE_TABLE_USER_NAME, $user->getName(), $newName);
+
+			// move user page
+			$oldUserPage = Title::newFromText($user->getName(), NS_USER);
+			$newUserPage = Title::newFromText($newName, NS_USER);
+			$oldUserPage->moveTo($newUserPage, false /* do not check permissions to move */, 'Migrating Uncyclopedia accounts');
+		}
+
+		// return the updated user object
+		$user->setName( $newName );
+		return $user;
+	}
+
+	/**
+	 * Remove given Uncyclopedia account
+	 *
+	 * @param User $user
+	 */
+	protected function deleteUncycloUser( User $user ) {
+		// delete the user and his settings from the uncyclo DB
+		$dbw = $this->getUncycloDB( DB_MASTER );
+
+		$dbw->delete( self::USER_TABLE,  [ 'user_id' => $user->getId() ], __METHOD__ );
+		$dbw->delete( 'user_properties', [ 'up_user' => $user->getId() ], __METHOD__ );
+	}
+
+	/**
+	 * Rename global user
+	 *
+	 * @param User $user global user
+	 */
+	protected function doRenameGlobalUser( User $user ) {
+		$newName = self::PREFIX_RENAME_GLOBAL . $user->getName();
+
+		// check user name conflicts
+		if ( $this->getGlobalUserByName( $newName ) ) {
+			$newName .= '-1';
+		}
+
+		$this->output( sprintf( '> renaming global user to "%s" (#%d)...', $newName, $user->getId() ) );
+
+		$this->renamedWikiaAccounts++;
+
+		if ( $this->isDryRun ) {
+			return;
+		}
+
+		$cmd = sprintf( 'php %s/renameUser.php --old-username=%s --new-username=%s --reason=%s',
+			__DIR__,
+			escapeshellarg( $user->getName() ),
+			escapeshellarg( $newName ),
+			escapeshellarg( self::RENAME_REASON )
+		);
+
+		$retVal = 0;
+		$output = wfShellExec( $cmd, $retVal );
```

Comment:
The user will be processed just once in the script, so there's no need to keep its state up to date after the background task(s) is completed.

More over, each user migration is wrapped in a transaction, so the state will be updated.